File: SunwooChoi/2022-09-15-Product-of-Array-Except-Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        products = [0] * len(nums)
        products[0] = 1
        acc = 1 # acculmulative product in reverse
        
        # product x[n] = 1 * x[0] * x[1] * ... * x[n-1]
        for idx in range(1, len(nums)):
            products[idx] = products[idx-1] * nums[idx-1]
        
        for idx in range(len(nums)-1, -1, -1):
            products[idx] = products[idx] * acc
            acc = acc * nums[idx]
        return products

File: DongwanKim/2022-09-15-ProductOfArrayExceptSelf.js
/**
 * Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.

 

Example 1:

Input: nums = [1,2,3,4]
Output: [24,12,8,6]
Example 2:

Input: nums = [-1,1,0,-3,3]
Output: [0,0,9,0,0]
 

Constraints:

2 <= nums.length <= 105
-30 <= nums[i] <= 30
The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.
 */

//first solution
var productExceptSelf = function(nums) {
    let answer = [];
    let pre = 1;
    for(let i=0; i<nums.length; i++){
        // put prefix
        answer[i] = pre;
        // update prefix
        pre *= nums[i]
    }
    let post = 1;
    for(let i=nums.length-1; i>=0; i--){
        // calculate with postfix
        answer[i] *= post;
        // update postfix
        post *= nums[i]
    }
    return answer;
};